src/realm/sync/subscriptions.cpp

```diff
@@ -859,18 +817,59 @@ void SubscriptionStore::terminate()
     }
 }
 
-MutableSubscriptionSet SubscriptionStore::get_mutable_by_version(int64_t version_id)
+void SubscriptionStore::update_state(int64_t version, State new_state, std::optional<std::string_view> error_str)
 {
+    REALM_ASSERT(error_str.has_value() == (new_state == State::Error));
+    REALM_ASSERT(new_state != State::Uncommitted);
+    REALM_ASSERT(new_state != State::Pending);
+    REALM_ASSERT(new_state != State::Superseded);
+
     auto tr = m_db->start_write();
     auto sub_sets = tr->get_table(m_sub_set_table);
-    auto obj = sub_sets->get_object_with_primary_key(Mixed{version_id});
+    auto obj = sub_sets->get_object_with_primary_key(version);
     if (!obj) {
-        throw KeyNotFound(util::format("Subscription set with version %1 not found", version_id));
+        // This can happen either due to a bug in the sync client or due to the
+        // server sending us an error message for an invalid query version. We
+        // assume it is the latter here.
+        throw RuntimeError(ErrorCodes::SyncProtocolInvariantFailed,
+                           util::format("Invalid state update for nonexistent query version %1", version));
     }
-    return MutableSubscriptionSet(weak_from_this(), std::move(tr), obj);
+
+    auto old_state = state_from_storage(obj.get<int64_t>(m_sub_set_state));
+    switch (new_state) {
+        case State::Error:
+            if (old_state == State::Complete) {
+                throw RuntimeError(
```

Comment:
I think this can happen if a queryable field is removed after the query is marked complete

Reply:
I'm not sure what we should do when that happens. The SubscriptionStore code assumes that the active subscription set is always valid and there isn't an obvious resolution when a query which used to be valid no longer is. This change is just adjusting the error message and code which is reported when it does happen.

Reply:
I'll add a note about this possibly due to a queryable field being removed to the error.